File: frontend/src/types/card.ts
export interface Card {
  id: string;
  name: string;
  mana_cost?: string;
  cmc: number;
  type_line: string;
  oracle_text?: string;
  power?: string;
  toughness?: string;
  colors?: string[];
  color_identity: string[];
  set: string;
  set_name: string;
  rarity: string;
  image_uris?: {
    small: string;
    normal: string;
    large: string;
    art_crop: string;
  };
  card_faces?: Array<{
    name: string;
    mana_cost?: string;
    type_line: string;
    oracle_text?: string;
    image_uris?: {
      small: string;
      normal: string;
      large: string;
      art_crop: string;
    };
  }>;
  prices: {
    usd?: string;
    usd_foil?: string;
  };
  legalities: Record<string, string>;
  scryfall_uri: string;
}

export interface SearchResult {
  object: string;
  total_cards: number;
  has_more: boolean;
  next_page?: string;
  data: Card[];
}

File: backend/models/card.py
from pydantic import BaseModel
from typing import Optional
from datetime import date


class CardBase(BaseModel):
    name: str
    mana_cost: Optional[str] = None
    cmc: float = 0.0
    type_line: str
    oracle_text: Optional[str] = None
    power: Optional[str] = None
    toughness: Optional[str] = None
    colors: list[str] = []
    color_identity: list[str] = []
    keywords: list[str] = []
    set_code: str
    rarity: str
    

class Card(CardBase):
    id: str
    scryfall_id: str
    oracle_id: Optional[str] = None
    image_uri: Optional[str] = None
    art_crop_uri: Optional[str] = None
    released_at: Optional[date] = None
    legalities: dict[str, str] = {}
    prices: dict[str, Optional[str]] = {}

    class Config:
        from_attributes = True


class CardCreate(CardBase):
    scryfall_id: str


class CardSearch(BaseModel):
    query: str
    colors: Optional[list[str]] = None
    types: Optional[list[str]] = None
    sets: Optional[list[str]] = None
    rarity: Optional[list[str]] = None
    cmc_min: Optional[float] = None
    cmc_max: Optional[float] = None
    limit: int = 20
    offset: int = 0
